pkg/controller/manager/nad/controller.go

```diff
@@ -148,27 +150,42 @@ func (h Handler) OnCNChange(_ string, cn *networkv1.ClusterNetwork) (*networkv1.
 	return nil, nil
 }
 
+// nad manager controller ensures all labels and sync cn
 func (h Handler) OnChange(_ string, nad *cniv1.NetworkAttachmentDefinition) (*cniv1.NetworkAttachmentDefinition, error) {
 	if nad == nil || nad.DeletionTimestamp != nil {
 		return nil, nil
 	}
 
 	klog.Infof("nad configuration %s/%s has been changed: %s", nad.Namespace, nad.Name, nad.Spec.Config)
 
-	if err := h.ensureLabels(nad); err != nil {
+	netconf, updated, err := h.ensureLabels(nad)
+	if err != nil {
 		return nil, fmt.Errorf("ensure labels of nad %s/%s failed, error: %w", nad.Namespace, nad.Name, err)
 	}
+	if updated {
+		// nad labels is updated by controller (rarely happens, it should be done via mutator)
+		// following IsVlanNad depends on those labels
+		// wait until it is updated and then go to below
+		return nad, nil
+	}
 
+	// overlay nad does not trigger following steps
+	if utils.IsOverlayNad(nad) {
+		return nad, nil
+	}
+
+	// the nad is not a vlan nad any more, always clear the job
 	if !utils.IsVlanNad(nad) {
 		if err := h.clearJob(nad); err != nil {
 			return nil, err
 		}
-		return nad, nil
+	} else {
+		if err := h.EnsureJob2GetLayer3NetworkInfo(nad, netconf); err != nil {
+			return nil, err
+		}
 	}
-
-	klog.Infof("nad configuration %s has been changed: %s", nad.Name, nad.Spec.Config)
-
-	if err := h.EnsureJob2GetLayer3NetworkInfo(nad); err != nil {
+	// nad change triggers the re-compute of cn's vlanset
+	if err := h.UpdateClusetNetworkVlanSet(nad); err != nil {
```

Comment:
would it not be better to just have a relatedresource watch in the cluster controller to watch NAD objects and trigger reconcile of cluster networks when a NAD changes? this would make it easier to follow the code.